x-pack/test/functional/services/machine_learning/data_visualizer_index_based.ts

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import expect from '@kbn/expect';
+
+import { FtrProviderContext } from '../../ftr_provider_context';
+
+export function MachineLearningDataVisualizerIndexBasedProvider({
+  getService,
+}: FtrProviderContext) {
+  const testSubjects = getService('testSubjects');
+  const retry = getService('retry');
+  const browser = getService('browser');
+
+  return {
+    async assertTimeRangeSelectorSectionExists() {
+      await testSubjects.existOrFail('mlDataVisualizerTimeRangeSelectorSection');
+    },
+
+    async assertTotalDocumentCount(expectedTotalDocCount: number) {
+      await retry.tryForTime(5000, async () => {
+        const docCount = await testSubjects.getVisibleText('mlDataVisualizerTotalDocCount');
+        expect(docCount).to.eql(expectedTotalDocCount);
+      });
+    },
+
+    async clickUseFullDataButton(expectedTotalDocCount: number) {
+      await testSubjects.clickWhenNotDisabled('mlButtonUseFullData');
+      await this.assertTotalDocumentCount(expectedTotalDocCount);
+    },
+
+    async assertFieldsPanelsExist(expectedPanelCount: number) {
+      const allPanels = await testSubjects.findAll('~mlDataVisualizerFieldsPanel');
+      expect(allPanels).to.have.length(2);
+    },
+
+    async assertFieldsPanelForTypesExist(fieldTypes: string[]) {
+      await testSubjects.existOrFail(`mlDataVisualizerFieldsPanel ${fieldTypes}`);
+    },
+
+    async assertCardExists(cardType: string, fieldName?: string) {
+      await testSubjects.existOrFail(`mlFieldDataCard ${fieldName} ${cardType}`);
+    },
+
+    async assertFieldsPanelCardCount(panelFieldTypes: string[], expectedCardCount: number) {
+      const filteredCards = await testSubjects.findAll(
+        `mlDataVisualizerFieldsPanel ${panelFieldTypes} > ~mlFieldDataCard`
+      );
+      expect(filteredCards).to.have.length(expectedCardCount);
```

Comment:
I would suggest to wrap this whole card count assertion in a `retry.tryForTime` to give it some time to load the cards e.g. after filtering.

Reply:
Done in https://github.com/elastic/kibana/pull/51832/commits/dfde35a493e596fe1db4d22b8d6dc00e0d14de52